Public/Src/Engine/Scheduler/Scheduler.cs

```diff
@@ -344,6 +344,26 @@ private void StartWorkers(LoggingContext loggingContext)
             m_allWorker = new AllWorker(m_workers.ToArray());
 
             ExecutionLog?.WorkerList(new WorkerListEventData { Workers = m_workers.SelectArray(w => w.Name) });
+        }
+
+        private readonly TaskSourceSlim<bool> m_schedulerCompletionExceptMaterializeOutputs = TaskSourceSlim.Create<bool>();
+        private bool AnyPendingPipsExceptMaterializeOutputs()
+        {
+            long numRunningOrQueued = m_pipQueue.NumRunningOrQueued;
+            long numRunningOrQueuedExceptMaterialize = numRunningOrQueued - m_pipQueue.GetNumRunningByKind(DispatcherKind.Materialize) - m_pipQueue.GetNumQueuedByKind(DispatcherKind.Materialize);

```

Comment:
>DispatcherKind.Materialize [](start = 107, length = 26)

nit: should we make this an argument to make the methods more 'generic'

Reply:
the if branch of the method is already related to MaterializeOutput part. Materialize step is pretty embedded in this method; so we do not need to make it generic.

---
In reply to: [348745577](https://github.com/microsoft/BuildXL/pull/1216#discussion_r348745577) [](ancestors = 348745577)